view/statistics-flow-certificates.js

```diff
@@ -0,0 +1,340 @@
+			return true;
+		}
+	});
+	floorToUnit(currentDate, mode.key);
+	while (currentDate <= dateTo) {
+		key         = mode.getDisplayedKey(currentDate);
+		result[key] = buildFilteredResult(data, key, service, certificate);
+		incrementDateByTimeUnit(currentDate, mode.key);
+	}
+
+	return result;
+};
+
+var calculateDurationByMode = function (dateFrom, dateTo, mode) {
+	var timeUnitsCount = 0, currentDate;
+	currentDate = copyDbDate(dateFrom);
+	if (!dateTo) dateTo = new db.Date();
+
+	floorToUnit(currentDate, mode);
+	while (currentDate <= dateTo) {
+		incrementDateByTimeUnit(currentDate, mode);
+		timeUnitsCount++;
+	}
+	return timeUnitsCount;
+};
+
+exports['statistics-main'] = function () {
+	var queryHandler, data = new ObservableValue({})
+	  , pagination = new Pagination('/flow/'), handlerConf;
+
+	handlerConf = queryHandlerConf.slice(0);
+	handlerConf.push({
+		name: 'page',
+		ensure: function (value, resolvedQuery, query) {
+			var num, dateFrom, dateTo, mode, durationInTimeUnits, resolvedValue;
+			resolvedValue = value;
+			if (resolvedValue == null) resolvedValue = '1';
+			if (isNaN(resolvedValue)) throw new Error("Unrecognized page value " + JSON.stringify(value));
+			num = Number(resolvedValue);
+			if (!isNaturalNumber(num)) throw new Error("Unreconized page value " + JSON.stringify(value));
+			if (num < 1) throw new Error("Unexpected page value " + JSON.stringify(value));
+
+			dateFrom = resolvedQuery.dateFrom;
+			dateTo   = resolvedQuery.dateTo || new db.Date();
+			mode     = resolvedQuery.mode;
+			durationInTimeUnits     = calculateDurationByMode(dateFrom, dateTo, mode);
+			resolvedQuery.pageCount = Math.ceil(durationInTimeUnits / itemsPerPage);
+			if (num > resolvedQuery.pageCount) throw new Error("Page value overflow");
+			return num;
+		}
+	}, serviceQuery, certificateQuery);
+	queryHandler = setupQueryHandler(handlerConf,
+		location, '/flow/');
+
+	queryHandler.on('query', function (query) {
+		var serverQuery = copy(query), dateFrom, dateTo, mode
+		  , currentDate, offset, timeUnitsCount = 0, durationInTimeUnits, now, page;
+
+		now      = new db.Date();
+		dateFrom = query.dateFrom;
+		dateTo   = query.dateTo || new db.Date();
+		mode     = query.mode;
+		page     = query.page;
+
+		durationInTimeUnits = calculateDurationByMode(dateFrom, dateTo, mode);
+		if (query.pageCount > 1) {
+			offset = { from: ((page - 1) * itemsPerPage) };
+			offset.to = offset.from;
+			if ((durationInTimeUnits - offset.from) < itemsPerPage) {
+				offset.to += durationInTimeUnits - offset.from;
+			} else {
+				offset.to += itemsPerPage;
+			}
+			offset.to -= 1;
+
+			currentDate = copyDbDate(dateFrom);
+			floorToUnit(currentDate, mode);
+			while (timeUnitsCount <= offset.to) {
+				if (timeUnitsCount === offset.from && query.page > 1) {
+					dateFrom = copyDbDate(currentDate);
+				}
+				if (timeUnitsCount === offset.to && query.page < query.pageCount) {
+					dateTo = copyDbDate(ceilToUnit(currentDate, mode));
```

Comment:
I think instead of introducing superfluous `ceilToUnit` we can just do:
```javascript
dateTo = incrementDateByTimeUnit(copyDbDate(currentDate), mode));
```

(there are also issues in `ceilToUnit`  but I didn't comment on them, as it's best if we remove it).

Additionally what's the agreement on `dateFrom` and `dateTo` handling? Are both inclusive?
If that's the case, then additionally we should:

```javascript
dateTo.setUTCDate(dateTo.getUTCDate() - 1);
```

(so e.g. instead of making `dateTo` the first day of the next month, we make it last day of current)